enterprise/reporting/error-index/error_index_reporting.go

```diff
@@ -0,0 +1,346 @@
+				EventType:        metric.StatusDetail.EventType,
+				ReceivedAt:       failedMessage.ReceivedAt,
+				FailedAt:         failedAt,
+			}
+			payloadJSON, err := json.Marshal(payload)
+			if err != nil {
+				return fmt.Errorf("unable to marshal payload: %v", err)
+			}
+
+			params := struct {
+				WorkspaceID string `json:"workspaceId"`
+				SourceID    string `json:"source_id"`
+			}{
+				WorkspaceID: workspaceID,
+				SourceID:    metric.SourceID,
+			}
+			paramsJSON, err := json.Marshal(params)
+			if err != nil {
+				return fmt.Errorf("unable to marshal params: %v", err)
+			}
+
+			jobs = append(jobs, &jobsdb.JobT{
+				UUID:         uuid.New(),
+				Parameters:   paramsJSON,
+				EventPayload: payloadJSON,
+				EventCount:   1,
+				WorkspaceId:  workspaceID,
+			})
+		}
+	}
+
+	if len(jobs) == 0 {
+		return nil
+	}
+	db, err := eir.resolveJobsDB(tx)
+	if err != nil {
+		return fmt.Errorf("failed to resolve jobsdb: %w", err)
+	}
+	if err := db.WithStoreSafeTxFromTx(eir.ctx, tx, func(tx jobsdb.StoreSafeTx) error {
+		return db.StoreInTx(eir.ctx, tx, jobs)
+	}); err != nil {
+		return fmt.Errorf("failed to store jobs: %v", err)
+	}
+
+	return nil
+}
+
+func (eir *ErrorIndexReporter) DatabaseSyncer(c types.SyncerConfig) types.ReportingSyncer {
+	eir.dbsMu.Lock()
+	defer eir.dbsMu.Unlock()
+
+	if _, ok := eir.dbs[c.ConnInfo]; ok {
+		return func() {} // returning a no-op syncer since another go routine has already started syncing
+	}
+
+	dbHandle, err := sql.Open("postgres", c.ConnInfo)
+	if err != nil {
+		panic(fmt.Errorf("failed to open error index db: %w", err))
+	}
+	errIndexDB := jobsdb.NewForReadWrite(
+		"err_idx",
+		jobsdb.WithDBHandle(dbHandle),
+		jobsdb.WithDSLimit(eir.conf.GetReloadableIntVar(0, 1, "Reporting.errorIndexReporting.dsLimit")),
+		jobsdb.WithConfig(eir.conf),
+		jobsdb.WithSkipMaintenanceErr(eir.conf.GetBool("Reporting.errorIndexReporting.skipMaintenanceError", false)),
+		jobsdb.WithJobMaxAge(
+			func() time.Duration {
+				return eir.conf.GetDurationVar(24, time.Hour, "Reporting.errorIndexReporting.jobRetention")
+			},
+		),
+	)
+	if err := errIndexDB.Start(); err != nil {
+		panic(fmt.Errorf("failed to start error index db: %w", err))
+	}
+	eir.dbs[c.ConnInfo] = &handleWithSqlDB{
+		Handle: errIndexDB,
+		sqlDB:  dbHandle,
+	}
+
+	if !eir.conf.GetBool("Reporting.errorIndexReporting.syncer.enabled", true) {
+		return func() {}
+	}
+
+	return func() {
+		eir.g.Go(func() error {
+			return eir.mainLoop(eir.ctx, errIndexDB)
+		})
+	}
+}
+
+func (eir *ErrorIndexReporter) mainLoop(ctx context.Context, errIndexDB *jobsdb.Handle) error {
+	eir.log.Infow("Starting main loop for error index reporting")
+
+	provider := eir.conf.GetString("JOBS_BACKUP_STORAGE_PROVIDER", "S3")
+	bucket := eir.conf.GetString("FAILED_JOBS_BACKUP_BUCKET", "rudder-failed-messages")
+	fm, err := filemanager.New(&filemanager.Settings{
+		Provider: provider,
+		Config: filemanager.GetProviderConfigFromEnv(
```

Comment:
We shouldn't assume default configuration for error index object storage. At least it should be possible to override it through custom configuration options.

Reply:
Do you mean we would need different config values for these?
```
JOBS_BACKUP_STORAGE_PROVIDER
AWS_ACCESS_KEY_ID
AWS_SECRET_ACCESS_KEY
AWS_ENABLE_SSE
AWS_S3_REGION_HINT
BACKUP_IAM_ROLE_ARN
```

Reply:
Do you mean to use something like this?

```GO
        provider := eir.conf.GetStringVar("S3", "ErrorIndex.Storage.Provider", "JOBS_BACKUP_STORAGE_PROVIDER")
	bucketName := eir.conf.GetStringVar("rudder-failed-messages", "ErrorIndex.Storage.Bucket")
	accessKeyID := eir.conf.GetStringVar("", "ErrorIndex.Storage.AccessKeyID", "AWS_ACCESS_KEY_ID")
	accessKey := eir.conf.GetStringVar("", "ErrorIndex.Storage.AccessKey", "AWS_SECRET_ACCESS_KEY")
	enableSSE := eir.conf.GetBoolVar(false, "ErrorIndex.Storage.EnableSSE", "AWS_ENABLE_SSE")
	region := eir.conf.GetStringVar("us-east-1", "ErrorIndex.Storage.RegionHint", "AWS_S3_REGION_HINT")
	iamRoleArn := eir.conf.GetStringVar("", "ErrorIndex.Storage.IAMRoleArn", "BACKUP_IAM_ROLE_ARN")

	fm, err := filemanager.New(&filemanager.Settings{
		Provider: provider,
		Config: map[string]interface{}{
			"bucketName":  bucketName,
			"accessKeyID": accessKeyID,
			"accessKey":   accessKey,
			"enableSSE":   enableSSE,
			"regionHint":  region,
			"iamRoleArn":  iamRoleArn,
		},
		Conf: eir.conf,
	})
	if err != nil {
		return fmt.Errorf("creating file manager: %w", err)
	}
```